### ISSUE TITLE
Bootstrap variables used by pkgdown

### ISSUE THREAD

**Issue by apreshill** [r-lib/pkgdown#1678]
Hi @maelle 👋 

I've been working with the dev version of pkgdown for a bit, and wanted to share what I learned converting a previous CSS to a hybrid/frankenstein BS4 variable setup with additional CSS. I ran into a few issues with Bootstrap 4 variables and tried to annotate as I went.

Here are the variables I used that did seem to work:

```
    bg: "white"                         # background of site
    primary: "#187BA2"                  # sidebar heading colors / but not content
    headings-color: "#187BA2"           # inherits from body if not set / not sidebars though
    navbar-bg: "#f4e4e7"                # navbar bg color / inherits primary if not set
    secondary: "#595959"                # version badge bg only?
    navbar-light-brand-color: "#595959" # pkgname color / inherits navbar-light-active-color if not set
```

And here are the variables that do not seem to work and require some brute-force CSS:

```
    # these don't work
    link-color: "#767676"               # should inherit primary, but doesn't? now pkgdown overrides
    link-hover-color: "#187BA2"         # should inherit and darken link-color, but pkgdown overrides
    fg: "#595959"                       # all body text, cannot leave blank? but also bootstrap overrides?
    navbar-light-active-color: "#187BA2"
    navbar-light-hover-color: "#126180" # pkgdown overrides
    navbar-light-color: "#595959"       # pkgdown overrides
```

And here was the CSS I needed:

```
/*----Pkgdown is overriding these----*/

/* fg in YAML */
body {
	color: var(--scorpion);
}

/* navbar-light-color in YAML */
.navbar-light .navbar-nav .nav-link {
	color: var(--scorpion);
}

/* navbar-light-hover-color in YAML */
.navbar-light .navbar-nav .nav-link:hover {
	color: var(--sapphire);
}

/* navbar-light-active-color in YAML */
.navbar-light .navbar-nav .active > .nav-link {
	color: var(--matisse);
}

/* link-hover-color */
a:hover {
  color: var(--sapphire);
}

/* link-color */
a {
  color: var(--boulder);
}
```

From looking at some of the commits and comments to the `pkgdown.css` file, it looks like some changes were made to improve accessibility (which is great!), but they seem to come at the cost of allowing users to use the Bootstrap variables. Anyway, I'm happy to funnel some of my notes above which variables affect what into your current vignette (https://pkgdown.r-lib.org/dev/articles/customization.html#bslib-variables), but was hoping you might have some ideas for making these 6 variables work 😄 

I'm really enjoying customizing this new template! Here is my repo for reference: https://github.com/apreshill/bakeoff

And the deployed site: https://bakeoff.netlify.app/

**Comment by maelle** [comment]
(so, thank you!!)

**Comment by maelle** [comment]
@apreshill the only variable I can't tackle is fg. Is it `body-color`?

**Comment by apreshill** [comment]
Ooh `body-color`! I'm just getting the hang of these global bs4 variables too- `fg` seems to be the bslib way of setting body color (https://rstudio.github.io/bslib/articles/bslib.html#main-colors), although I don't see it documented anywhere else on Bootstrap docs. This is the `flexdashboard` theming vignette for more context: https://pkgs.rstudio.com/flexdashboard/articles/theme.html#custom-themes

Will do PR today 👩‍🍳 Are you open to an R-based interactive table (i.e., a reactable, a DT like bslib), or prefer plain markdown? I think it would have helped me if I could have seen:

+ the variable name (selecting likely these 12 as the most relevant for pkgdown sites)
+ what it would inherit from if not set (if anything)
+ narrative text description

Also- I'm not sure if you noticed, but `secondary` seems largely unused- only for the background of the version badge AFAIK (I've just realized I did not attempt footer elements, so it may cascade there). I was wondering if that variable might be made more useful? One idea I had (b/c it was difficult to tweak with CSS) was to make secondary

**Comment by maelle** [comment]
Thanks so much @apreshill, super useful. Regarding CSS changes before the docs improvements,

* New BS variables in #1682 (so one could use bslib variables to override these defaults)
* If I follow correctly pkgdown.css should also be edited to use more BS variables in particular `secondary`?

Some notes 
* the footer rules could use variables instead https://github.com/r-lib/pkgdown/blob/71849077a7232b20d0f76eb7203e558b437e0dd7/inst/assets/BS4/pkgdown.css#L38
* https://github.com/r-lib/pkgdown/blob/71849077a7232b20d0f76eb7203e558b437e0dd7/inst/assets/BS4/pkgdown.css#L140, https://github.com/r-lib/pkgdown/blob/71849077a7232b20d0f76eb7203e558b437e0dd7/inst/assets/BS4/pkgdown.css#L146
* the Mark color for highlighting of search in result pages? https://github.com/r-lib/pkgdown/blob/71849077a7232b20d0f76eb7203e558b437e0dd7/inst/assets/BS4/pkgdown.css#L239
* colors for search results https://github.com/r-lib/pkgdown/blob/71849077a7232b20d0f76eb7203e558b437e0dd7/inst/assets/BS4/pkgdown.css#L405
* badge code, as noted could become something else but the change would be in HTML & R code, not CSS https://github.com/r-lib/pkgdown/blob/71849077a7232b20d0f76eb7203e558b437e0dd7/inst/templates/BS4/navbar.html#L10

I suppose making these changes will also make creating the dark mode much easier. I could look into making changes later this week.

Also related https://github.com/rstudio/bookdown/issues/1021 

The syntax highlighting CSS would remain separate I suppose. :thinking:

**Comment by apreshill** [comment]
Hi- sorry for the pause, I wanted to take some time yesterday to dig a bit deeper into both Bootstrap cascades, and how the Bootswatch themes make use of variables and the cascade too. I did find an interesting candidate variable that we could map `$secondary` onto: `$component-active-bg`. Seen here:

<img width="1300" alt="Screen Shot 2021-05-13 at 11 38 28 AM" src="https://user-images.githubusercontent.com/12160301/118171083-c50d6380-b3df-11eb-853e-d13fe23085b8.png">


If we let `$secondary` map onto this variable, I think we should make it affect four elements (see the first reactable table here; there are 3 total: one for theme variables, one for overriding theme variables, and one for navbar light/dark variables). 

https://focused-snyder-9f2e47.netlify.app/articles/customization.html#bslib-variables-1

What do you both think? (I made these tables more for us- this is not intended as a draft of user-facing documentation-yet 😄 )

**Comment by maelle** [comment]
Thanks again @apreshill! I'm digesting all your feedback into the open PR #1682 

Regarding the version badge, after experimenting I think the current behavior makes sense: the badge is "danger" for the in development website but "default" (hence secondary-colored) for the release version.

**Comment by apreshill** [comment]
I agree, "danger" definitely makes sense for in development- can you override the default cascade from `$secondary` if you use a more specific variable?

I'm also still digesting as well! I'll continue getting some feedback internally about what variable cascades would be least surprising 🤗

**Comment by maelle** [comment]
Btw @apreshill when you write " I did find an interesting candidate variable that we could map $secondary onto: $component-active-bg" do you have any example of how to do that with sass? I'm trying to use grey-200 for secondary, and am failing to do so.

**Comment by maelle** [comment]
> I agree, "danger" definitely makes sense for in development- can you override the default cascade from $secondary if you use a more specific variable?

It didn't seem so so in my PR I now use badge-info for the release mode. It's `$cyan` by default, but one can tweak `$info` to get a different color just for the badge.

**Comment by maelle** [comment]
@apreshill actually fg works? I've created a monster with red fg and grey bg via the config:

![image](https://user-images.githubusercontent.com/8360597/118290606-ea30cd80-b4d6-11eb-93d6-3641782b2b01.png)

**Comment by maelle** [comment]
One last thing, reg the navbar the default is light at the moment

https://github.com/r-lib/pkgdown/blob/48d88cbb6dcfb10eb3ae5c6abec0cb7af5249e5f/R/navbar.R#L32-L33

**Comment by apreshill** [comment]
> Btw @apreshill when you write " I did find an interesting candidate variable that we could map $secondary onto: $component-active-bg" do you have any example of how to do that with sass? I'm trying to use grey-200 for secondary, and am failing to do so.

This is likely an oversimplification, as I've only used Sass via Hugo pipes. What I was thinking was that you could first do something like:

```
#pseudo-sass code
.navbar-light .navbar-nav .active > .nav-link {
	@extend #{$component-active-bg}; 
}
```

I did something sort of like this here: https://github.com/hugo-apero/hugo-apero/blob/05497e6e4382f861c02a52fbdcbd1bdc60cee236/assets/scaffold.scss#L475-L487

This would force the component we want (the active background color for a nav link) "listen to" the variable `$component-active-bg`.

Then, to make sure the that variable can be set with `$secondary`...

```
#pseudo-sass code
$component-active-bg: '$secondary' !default;
```

Again, this is probably oversimplified, but was just working out an existing Bootstrap variable that might make sense!

**Comment by maelle** [comment]
Thank you! At the moment I'm trying, before running `bslib::bs_theme_dependencies()`,

```r
# map secondary to component-active-bg
secondary <- bslib::bs_get_variables(bs_theme, "secondary")
bs_theme <- bslib::bs_add_variables(
bs_theme,
"component-active-bg" = as.character(secondary)
)
```

which doesn't work yet.

**Comment by maelle** [comment]
Actually it does work for the button!

**Comment by maelle** [comment]
I had to keep the rules with `$component-active-bg` in them, I wonder whether I missed something though I'm not too surprised regarding the rules of the TOC.

**Comment by maelle** [comment]
mmmh it was simpler actually. :woman_facepalming:

**Comment by apreshill** [comment]
just waking up and excited to test your new branch out! 👯

**Comment by maelle** [comment]
wait a bit then :see_no_evil: 

Surprised this does not work hence my workaround:

``` r
library("bslib")
theme <- bs_theme(
   "primary" = "#fff",
    "component-active-bg" = "$primary  !default"
  )
bslib::bs_theme_dependencies(theme)
#> Error in compile_data(sass_input, options): Error: Undefined variable: "$primary".
#>         on line 2:23 of stdin
#> >> $component-active-bg: $primary  !default;
#> 
#>    ----------------------^
```

<sup>Created on 2021-05-20 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0.9000)</sup>

**Comment by maelle** [comment]
ok now ready for testing @apreshill

**Comment by maelle** [comment]
- [ ] Can Alison now use the variables that should work (from the first comment) https://github.com/r-lib/pkgdown/issues/1678#issue-885426843
- [ ] Why is my code for mapping secondary to component-active-bg so clunky cf https://github.com/r-lib/pkgdown/issues/1678#issuecomment-845153696
- [ ] Once my PR is merged i.e. we all agree on the variables, adding documentation.

Am I missing something?

**Comment by maelle** [comment]
Current workflow for replacing values in pkgdown.sass:

* find whether there's a logical BS variable, use it;
* in render.R in pkgdown defaults add a default value for an ancestor of that variable (as little specific as possible). E.g. [`5b5fdff` (#1682)](https://github.com/r-lib/pkgdown/pull/1682/commits/5b5fdff5ad3ceb38e353de252748445b36ac111a)

Does this make sense? (i.e. have I learnt my lesson? :sweat_smile: )

**Comment by apreshill** [comment]
Starting off small and working my way up:

If I just use (woot!):

```
fg: "#982649" 
bg: "white"
```

The footer color doesn't inherit `fg`, maybe should?:

```
footer {
	color: #666;
}
```

<img width="1257" alt="Screen Shot 2021-05-20 at 3 59 54 PM" src="https://user-images.githubusercontent.com/12160301/119059106-826dfd00-b984-11eb-87f1-1167622bdbe0.png">

It looks like a nice alpha is added to `fg` by default 💅 Is that right?

---

Adding `primary`:

```
fg: "#ff0000" 
bg: "white"
primary: "#9B5094"   
```

Changed:

- [x] `link-color`

- [x] `link-hover-color` (darken `link-color` by 15% 😄 )

- [x] The background hover/active copy button and the border upon hover

- [ ] And maybe other vars too? [Maybe though - just a proposal!](https://focused-snyder-9f2e47.netlify.app/articles/customization.html#bslib-variables-1) Maybe not- this may be too much 👼 

---

Adding `secondary`:

```
bg: "white"                     
fg: "#982649"                    
primary: "#60B2E5"         
secondary: "#FBD989" 
```

oh WOW- all that CSS pain taken away! This is lovely. Also changed background color for copy button when focus/active.

<img width="1257" alt="Screen Shot 2021-05-20 at 4 25 10 PM" src="https://user-images.githubusercontent.com/12160301/119060885-f8c02e80-b987-11eb-9391-0a3b056f877a.png">

**Narrator's voice** After making more changes to `navbar-bg`, I did notice that the hover background color for dropdowns can get lost, maybe both hover and focus of dropdowns should track with `secondary` too?

<img width="1257" alt="Screen Shot 2021-05-20 at 4 56 02 PM" src="https://user-images.githubusercontent.com/12160301/119062791-6e2dfe00-b98c-11eb-84c0-b7213e21205d.png">
 

---

Adding `navbar-bg`

```
bg: "white"
fg: "#982649"
primary: "#60B2E5"
secondary: "#FBD989"
navbar-bg: "#F7CCB6"                 # must set bg key in navbar
```

Works if I also set bg key in the navbar:

<img width="1257" alt="Screen Shot 2021-05-20 at 4 34 11 PM" src="https://user-images.githubusercontent.com/12160301/119061435-3a9da480-b989-11eb-9b59-773c373dab33.png">


---

Adding `heading-color` works ✔️ 

```
bg: "white"
fg: "#982649"
primary: "#60B2E5"
secondary: "#FBD989"
navbar-bg: "#F7CCB6"                 # must set bg key in navbar
headings-color: "#006DAA"         # inherits from fg if not set
```

<img width="1257" alt="Screen Shot 2021-05-20 at 4 40 16 PM" src="https://user-images.githubusercontent.com/12160301/119061854-2dcd8080-b98a-11eb-8b3e-f4601343c717.png">


---

Moving more quickly now as everything works! All the link coloring works 🎉 

side note: adding `body-color` still works at this point! Still doesn't affect footer text color though.

I cannot seem to get `component-active-color` to affect anything- I was expecting this to change the text color anywhere that `component-active-bg` affected the background color (so text of active navbar links, text of active sidebar TOC links, and focus icon color for copy/paste button?)

```
bg: "white"
fg: "#982649"
primary: "#60B2E5"
secondary: "#FBD989"
navbar-bg: "#F7CCB6"                            # must set bg key in navbar
headings-color: "#006DAA"                    # inherits from fg if not set
link-color: "#B33C86"                             # inherit primary if not set
link-hover-color: "#EA638C"                  # darkens primary if not set
component-active-bg: "#AADACD"       # yes!
component-active-color: "#9FFFF5"     # nope
```

OK going to start hitting navbar with testing, but going to start new comment for that!!

**Comment by apreshill** [comment]
Navbar! Here we go...

Starting from:

```
    bg: "white"
    fg: "#982649"
    primary: "#60B2E5"
    secondary: "#FBD989"
    navbar-bg: "#F7CCB6"                 # must set bg key in navbar
    headings-color: "#006DAA"            # inherits from fg if not set
    link-color: "#B33C86"                # inherit primary if not set
    link-hover-color: "#EA638C"          # darkens primary if not set
    component-active-bg: "#AADACD"       # yes!
    component-active-color: "#9FFFF5"    # nope
```

OK everything light that I tried works! I tested:

```
    # navbar
    navbar-light-brand-color: "#476A6F"  # pkgname color / inherits navbar-light-active-color if not set
    navbar-light-brand-hover-color: "#B7094C" # pkgname hover / inherits navbar-light-hover-color if not set
    navbar-light-active-color: "#B7094C" # active link
    navbar-light-hover-color: "#1B998B"  # hover link
    navbar-light-color: "#2A9D8F"       # link
    info: "#EE92C2"
```

<img width="1257" alt="Screen Shot 2021-05-20 at 8 28 57 PM" src="https://user-images.githubusercontent.com/12160301/119077906-071f4200-b9aa-11eb-889a-709a7e29377d.png">


I'll pare down my reactable table for the vignette based on this. What do you think?

**Comment by apreshill** [comment]
Pulling out some notes as a tl;dr:

- [ ] what do you think about letting `component-active-color` affect all the text for elements that `component-active-bg` affects now? 
- [ ] what do you think about `primary` -> `navbar-light-active-color` + `component-active-color`? Too much?
- [ ] what do you think about the copy to clipboard button border always matching the background color? There aren't any other borders really, so the outline may be unnecessary
- [ ] what do you think about having the dropdown hover/focus follow `secondary` if set?
- [ ] should footer text color follow `fg` and then `body-color` if set?
- [ ] would it possible to use `navbar-bg` without setting `bg: navbar-bg` in the `navbar` section?

**Comment by maelle** [comment]
Thanks a ton for all of this!

Regarding the footer I've now gone with `color: mix($body-color, $body-bg, 96%);` which means it'll automatically be lighter than normal text. The percentage could be tweaked if someone disagrees.

**Comment by maelle** [comment]
> It looks like a nice alpha is added to fg by default nail_care Is that right?

Not by pkgdown.sass in any case. :thinking:

**Comment by maelle** [comment]
I've now made it possible to set the navbar bg via navbar-bg too.

**Comment by maelle** [comment]
> I cannot seem to get component-active-color to affect anything- I was expecting this to change the text color anywhere that component-active-bg affected the background color (so text of active navbar links, text of active sidebar TOC links, and focus icon color for copy/paste button?)

pkgdown.sass fixed for this! Good catch!

**Comment by maelle** [comment]
> what do you think about primary -> navbar-light-active-color + component-active-color? Too much?

I'd first like to have an answer on https://github.com/rstudio/bslib/issues/317 before I add more clunky code. :sweat_smile: would you mind opening a separate issue on this specific mapping? thank you!

**Comment by maelle** [comment]
> what do you think about having the dropdown hover/focus follow secondary if set? (dropdown variables here: rstudio/bslib@e0503c6/inst/lib/bs/scss/_variables.scss#L781)

Could you also open an issue about this? I think it's a good idea but I'd like to add it in a tidy way.

**Comment by maelle** [comment]
Reg docs I think it'll be better to wait until @hadley reviews the PR, to not document a moving target. Thanks again for all the extremely useful hints & tips!

**Comment by maelle** [comment]
>  what do you think about primary -> navbar-light-active-color + component-active-color? Too much?

@apreshill this is now done thanks to @cderv :tada:

**Comment by maelle** [comment]
Last box checked :tada:

**Comment by maelle** [comment]
Now border-color too is by default a mix of body-bg and body-color.

**Comment by maelle** [comment]
A missing mapping is probably the color of text in the navbar. :thinking: 

![image](https://user-images.githubusercontent.com/8360597/119149572-ef00fe80-ba4d-11eb-8102-3d62ff12d534.png)

**Comment by apreshill** [comment]
Yes- you mean the non-active/non-hover link color text? With bootswatch, it looks like it is white if dark: https://patchwork.data-imaginist.com/ 

Looking at bslib variable list I see:

| variable              | value                                                          | comment                                                                                                                        |
|-----------------------|----------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------|
| navbar-light-contrast | if($navbar-light-bg, color-contrast($navbar-light-bg), $black) | Deepest contrasting color for navbarPage(inverse = FALSE). Defaults to $white or $black based on luminance of $navbar-light-bg |
| navbar-light-color    | rgba($navbar-light-contrast, .5)                               |                                                                                                                                |

Looks like I cannot currently change anything with `navbar-light-contrast`. If set, it looks like it provides a default for hover/active colors too?
<img width="1257" alt="Screen Shot 2021-05-21 at 7 15 42 AM" src="https://user-images.githubusercontent.com/12160301/119151605-69a52c00-ba04-11eb-8674-35fdf15bc7ec.png">

**Comment by maelle** [comment]
@apreshill we've made `component-active-bg` inherit from `secondary` instead of `primary` and now I'm wondering whether this is bad for using Bootswatch themes. :thinking:

**Comment by maelle** [comment]
Reg `navbar-light-contrast` we can't use it because of

* the override in pkgdown.sass

```sass
.navbar-light .navbar-nav .active > .nav-link {
   background: $component-active-bg;
   color: $component-active-color;
   border-radius: 1rem;
 }
```

* the default in pkgdown's list of bslib variables on `navbar-light-color` and `navbar-light-hover-color` that I don't want to remove as the default would be much too light (`rgba($navbar-light-contrast, .5)`).

I've however changed the default: now if `navbar-light-color` and `navbar-light-hover-color` aren't set in the configuration, they're a bit lighter than the body color. Does that seem correct? This still means `navbar-light-contrast` is not usable.

**Comment by maelle** [comment]
Just some notes for myself. The different places where we are setting/changing things.
- Bootstrap has variables giving values to some things e.g. `$primary` for link-color, and inheritance rules for variables e.g. `$component-active-bg` inherits from `$primary`. Variables also have default values (or not). This is all seen in https://rstudio.github.io/bslib/articles/bs4-variables.html.
- We are setting default values to some variables i.e. actual hex codes and such in render.R, e.g. `$primary` to `#0054AD`.
- We are adding variables in rules in pkgdown.sass e.g.

```sass
.btn-copy-ex:active:hover {
  background-color: $component-active-bg;
  color: $component-active-color;
  border-color: $component-active-bg;
}
```

- We are tweaking inheritance rules in render.R e.g. `$component-active-bg` inheriting from secondary instead of primary.
- The users can add values for variables. (they can also add CSS but hopefully don't need to). So for instance `$component-active-bg` only inherits from `$secondary` if `$component-active-bg` is not set in the configuration by the user.
